src/encoding.rs

```diff
@@ -178,6 +178,80 @@ where
     Err(DecodeError::new("invalid varint"))
 }
 
+/// Additional information passed to every decode/merge function.
+///
+/// The context should be passed by value and can be freely cloned. When passing
+/// to a function which is decoding a nested object, then use `enter_recursion`.
+#[derive(Clone, Debug)]
+pub struct DecodeContext {
+    #[cfg(not(feature = "no-recursion-limit"))]
+    /// How many times we can recurse in the current decode stack before we hit
+    /// the recursion limit.
+    ///
+    /// The recursion limit is defined by `RECURSION_LIMIT` and cannot be
+    /// customized. The recursion limit can be ignored by building the Prost
+    /// crate with the `no-recursion-limit` feature (which is set by default).
+    recurse_count: u32,
+}
+
+impl Default for DecodeContext {
+    #[cfg(not(feature = "no-recursion-limit"))]
+    #[inline(always)]
```

Comment:
`inline` is a fairly weak hint to the compiler, `inline(always)` is stronger and I wanted to ensure that adding the context does not add any overhead if we disable the recursion limit.

Reply:
I don't think it actually matters - the compiler is likely to inline such a trivial function

Reply:
I ask because everything piece of advice I've seen on the subject recommends to use `#[inline]`, and not `#[inline(always)]`.  Anecdotally I've found that `#[inline(always)]` doesn't usually help.